src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/Shake256.cs

```diff
@@ -4,35 +4,196 @@
+
+                if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                {
+                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                }
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _concurrentCount);
+            }
+        }
+
+        internal unsafe void GetHashAndReset(Span<byte> destination)
+        {
+            int count = Interlocked.Increment(ref _concurrentCount);
+
+            try
+            {
+                if (count != 1)
+                {
+                    throw new CryptographicException(SR.Cryptography_ConcurrentUseNotSupported);
+                }
+
+                fixed (byte* pDestination = &Helpers.GetNonNullPinnableReference(destination))
+                {
+                    NTSTATUS ntStatus = Interop.BCrypt.BCryptFinishHash(_hashHandle, pDestination, destination.Length, dwFlags: 0);
+
+                    if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                    {
+                        throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                    }
+                }
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _concurrentCount);
+            }
+        }
+
+        internal void GetCurrentHash(Span<byte> destination)
+        {
+            // Doesn't need a concurrency blocker, because clone has one, and this
+            // doesn't expose the instance.
+            using (Shake256 clone = Clone())
+            {
+                clone.GetHashAndReset(destination);
+            }
+        }
+
+        internal Shake256 Clone()
+        {
+            int count = Interlocked.Increment(ref _concurrentCount);
+
+            try
+            {
+                if (count != 1)
+                {
+                    throw new CryptographicException(SR.Cryptography_ConcurrentUseNotSupported);
+                }
+
+                SafeBCryptHashHandle clone = Interop.BCrypt.BCryptDuplicateHash(_hashHandle);
+                return new Shake256(clone);
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _concurrentCount);
+            }
+        }
+
+        public void Dispose()
+        {
+            _hashHandle?.Dispose();
+        }
+
+        private static SafeBCryptHashHandle CreateHashHandle()
+        {
+            NTSTATUS ntStatus = Interop.BCrypt.BCryptCreateHash(
+                (nuint)Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_CSHAKE256_ALG_HANDLE,
+                out SafeBCryptHashHandle hashHandle,
+                pbHashObject: IntPtr.Zero,
+                cbHashObject: 0,
+                secret: ReadOnlySpan<byte>.Empty,
+                cbSecret: 0,
+                Interop.BCrypt.BCryptCreateHashFlags.None);
+
+            if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+            {
+                hashHandle.Dispose();
```

Comment:
hashHandle?.Dispose();

Reply:
I believe it's impossible for `hashHandle` to be null there (P/Invoke stubs and LibraryImport generated wrappers have to `new`-up the handle then use SetHandle on it, otherwise there's the possibility that the native method returns a thing to track, but .NET hits an OOM creating the SafeHandle... then nothing is around to clean up the native resource.

But, I can add the `?`, it doesn't cost much :smile: